### ISSUE TITLE
SSH login delayed by 25 seconds with pam_systemd(sshd:session): Failed to create session: Connection timed out

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#30348]
## Issue description

On NixOS 17.09 machines, initial SSH connections after boot are delayed for exactly 25 seconds.

I think this is a regression from 17.03, I didn't observe it there.

`journalctl --boot --unit sshd`

```
Oct 12 12:20:22.652948 gluster-cluster-1 sshd[726]: Server listening on 0.0.0.0 port 22.
Oct 12 12:20:22.653218 gluster-cluster-1 sshd[726]: Server listening on :: port 22.
Oct 12 12:20:22.995019 gluster-cluster-1 sshd[766]: Accepted publickey for root from 1.2.3.4 port 52664 ssh2: RSA SHA256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
Oct 12 12:20:22.997136 gluster-cluster-1 sshd[766]: pam_unix(sshd:session): session opened for user root by (uid=0)
# 25 seconds pass
Oct 12 12:20:48.022631 gluster-cluster-1 sshd[766]: pam_systemd(sshd:session): Failed to create session: Connection timed out
Oct 12 12:20:51.050106 gluster-cluster-1 sshd[1238]: Accepted publickey for root from 1.2.3.4 port 52266 ssh2: RSA SHA256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

The error here is `pam_systemd(sshd:session): Failed to create session: Connection timed out`.

This issue has been observed in Debian in two places but results were inconclusive:

* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=793814
* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=770135

And a systemd bug report that also doesn't really suggest what to do:

* https://github.com/systemd/systemd/issues/2863

There is some talk about dbus in it though, and I noticed that on my machines:

```
# systemctl list-unit-files --user
UNIT FILE                    STATE   
dbus.service                 indirect
systemd-exit.service         static  
dbus.socket                  disabled
basic.target                 static  
bluetooth.target             static  
busnames.target              static  
default.target               static  
exit.target                  static  
graphical-session-pre.target static  
graphical-session.target     static  
paths.target                 static  
printer.target               static  
shutdown.target              static  
smartcard.target             static  
sockets.target               static  
sound.target                 static  
timers.target                static  

17 unit files listed.
```

Note that dbus is `disabled` (and red in my terminal). However, it is the same on my 17.03 machine where I don't have this problem.

I've already tried `UseDNS no` for sshd but that doesn't change anything.

## Technical details

* System: NixOS deployed with NixOps (nixpkgs commit below)
* Nixpkgs version: `release-17.09` (commit `dea58ce228e6f9686773745024eb844d8dc39fc3`)

**Comment by nh2** [comment]
Possibly related?

* [systemd-logind must be restarted every ~1000 SSH logins to prevent a ~25 second delay](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1591411)
* [dbus bug Bug 95263 - unix fd in-flight counting broken](https://bugs.freedesktop.org/show_bug.cgi?id=95263) with a fix in `dbus 1.11.10` (I use 1.10.*)

**Comment by devurandom** [comment]
systemd/systemd#2863 was fixed by systemd/systemd#7102. Does that also fix this issue?

**Comment by nh2** [comment]
@devurandom I had a suspicion ([see here](https://github.com/systemd/systemd/pull/7102#issuecomment-338481611)) that https://github.com/systemd/systemd/pull/7102 might the symptom but not the cause.

The train of thought is the following:

* There is a difference between the behaviour of the fstab generator and mount units that we're not aware of that is possibly the key bug; this difference did not exist in an older systemd version
* `pam_systemd(login:session): Failed to create session: Connection timed out` appearing was a nice repro to observe that difference
* https://github.com/systemd/systemd/pull/7102 might make that error go away (I haven't had a chance to test that yet), thus possibly removing my chance to repro but the underlying problematic difference between fstab processing and mount unit behaviour might still exist

But I'm not very sure about it.

**Comment by volth** [comment]
perhaps different, but very similar bug:

25 seconds delay and ```pam_systemd(slim:session): Failed to create session: Connection timed out``` in system log take place on each startup on the virtual machines which have no network interfaces except ```lo```

reproducible on current (2018-04-05) master and probably on release-18.03

**Comment by r-darwish** [comment]
I'm not using NixOS nor Nix, but I had the same issue on Arch Linux when systemd 238 came. Downgrading back to systemd 237 solved it.

**Comment by volth** [comment]
It is with 237

**Comment by MasseGuillaume** [comment]
I get this error when I boot my computer. There is a 25 seconds delay. This is kind of annoying. I'm on 19.09pre184855.beff2f8d75e (Loris).

**Comment by nikitashalnov** [comment]
Faced with this issue as well, but don't use NixOS (debian instead it). `systemctl restart systemd-logind` helped to resolve this issue

**Comment by nh2** [comment]
still important to me

**Comment by nikrozman21** [comment]
Getting the exact same issue, but on CentOS 7.
Systemd 219

**Comment by gleventhal** [comment]
It's the dbus timeout.  It's 25 seconds.  You can reduce it using a limit stanza under /etc/dbus* and SSH will login faster when the dbus is under load.

**Comment by jas8522** [comment]
Using CentOS 8.3, I was encountering this issue on multiple servers. A totally unrelated issue had me install polkit to repair that glitch. Given that polkit regulates user permissions and this delay is connected to systemd-logind, I thought perhaps there's a connection and installed polkit on each of the servers with this issue and the problem is now fixed. Hope this might help others.

**Comment by jas8522** [comment]
> Did your systemd version or dbus version change as a result of polkit dependencies? I thought polkit was just that javascript-driven auth API, it seems odd that its presence would resolve the issue as I understand it.

No, they had been previously updated at least a few days before. Granted the server hadn't been restarted (though I had tried restarting logind and a systemd daemon reload, which did not fix it). I suppose it's plausible that the installation of polkit forced the restart of some other service that was contributing to the issue. If so, I wish I knew which one it was...

**Comment by MetalArend** [comment]
I keep restarting my laptop with NixOS a few times, because it starts up to a tty1 login screen, instead of a gui screen to login. The error happens with the gdm-launch-environment: 
```
Apr 21 10:18:39 nixos gdm-launch-environment][1035]: pam_systemd(gdm-launch-environment:session): Failed to create session: Connection timed out
Apr 21 10:18:39 nixos /nix/store/iwvnb7wzrjvqrz4k05hpjlhbrfqdqq72-gdm-40.0/libexec/gdm-wayland-session[1225]: Unable to run session message bus
```

**Comment by oluceps** [comment]
Encountered too. After I adding a systemd mount unit:
```
[Unit]
After=var-cache-jfs.mount minio.service
Description=Juicefs

[Mount]
Environment=AWS_DEFAULT_REGION=ap-east-1
Options=_netdev,allow_other,writeback_cache
Type=juicefs
What=postgres://juice:3b@/juicemeta?host=/var/run/postgresql/
Where=/share


[Install]
WantedBy=remote-fs.target multi-user.target
```
Then I need to waitting for 25s to enter machine. And the mount unit itself seems works well.
